provers/stark/src/examples/read_only_memory.rs

```diff
@@ -0,0 +1,435 @@
+        let alpha = &rap_challenges[1];
+        let a1 = second_step.get_main_evaluation_element(0, 0);
+        let v1 = second_step.get_main_evaluation_element(0, 1);
+        let a_sorted_1 = second_step.get_main_evaluation_element(0, 2);
+        let v_sorted_1 = second_step.get_main_evaluation_element(0, 3);
+        // (z - (a'_{i+1} + α * v'_{i+1})) * p_{i+1} = (z - (a_{i+1} + α * v_{i+1})) * p_i
+        let res = (z - (a_sorted_1 + alpha * v_sorted_1)) * p1 - (z - (a1 + alpha * v1)) * p0;
+
+        // The eval always exists, except if the constraint idx were incorrectly defined.
+        if let Some(eval) = transition_evaluations.get_mut(self.constraint_idx()) {
+            *eval = res;
+        }
+    }
+}
+
+pub struct ReadOnlyRAP<F>
+where
+    F: IsFFTField,
+{
+    context: AirContext,
+    trace_length: usize,
+    pub_inputs: ReadOnlyPublicInputs<F>,
+    transition_constraints: Vec<Box<dyn TransitionConstraint<F, F>>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct ReadOnlyPublicInputs<F>
+where
+    F: IsFFTField,
+{
+    pub a0: FieldElement<F>,
+    pub v0: FieldElement<F>,
+    pub a_sorted0: FieldElement<F>,
+    pub v_sorted0: FieldElement<F>,
+}
+
+impl<F> AIR for ReadOnlyRAP<F>
+where
+    F: IsFFTField + Send + Sync + 'static,
+    FieldElement<F>: ByteConversion,
+{
+    type Field = F;
+    type FieldExtension = F;
+    type PublicInputs = ReadOnlyPublicInputs<F>;
+
+    const STEP_SIZE: usize = 1;
+
+    fn new(
+        trace_length: usize,
+        pub_inputs: &Self::PublicInputs,
+        proof_options: &ProofOptions,
+    ) -> Self {
+        let transition_constraints: Vec<
+            Box<dyn TransitionConstraint<Self::Field, Self::FieldExtension>>,
+        > = vec![
+            Box::new(ContinuityConstraint::new()),
+            Box::new(SingleValueConstraint::new()),
+            Box::new(PermutationConstraint::new()),
+        ];
+
+        let context = AirContext {
+            proof_options: proof_options.clone(),
+            trace_columns: 5,
+            transition_offsets: vec![0, 1],
+            num_transition_constraints: transition_constraints.len(),
+        };
+
+        Self {
+            context,
+            trace_length,
+            pub_inputs: pub_inputs.clone(),
+            transition_constraints,
+        }
+    }
+
+    fn build_auxiliary_trace(
+        &self,
+        trace: &mut TraceTable<Self::Field, Self::FieldExtension>,
+        challenges: &[FieldElement<F>],
+    ) {
+        let main_segment_cols = trace.columns_main();
+        let a = &main_segment_cols[0];
+        let v = &main_segment_cols[1];
+        let a_sorted = &main_segment_cols[2];
+        let v_sorted = &main_segment_cols[3];
+        let z = &challenges[0];
+        let alpha = &challenges[1];
+
+        let trace_len = trace.num_rows();
+
+        let mut aux_col = Vec::new();
+        let num = z - (&a[0] + alpha * &v[0]);
+        let den = z - (&a_sorted[0] + alpha * &v_sorted[0]);
```

Comment:
For learning purposes this is ok, but for performance reasons, it's best to compute all the denominators, due batch inverse and then compute